New feature: inactive top cells

cbegeman opened the pull request:
Inactive top cells are introduced to add flexibility to the vertical coordinate below ice shelves. This PR introduces phase 1 in the development documented in https://github.com/MPAS-Dev/MPAS-Model/pull/811, which touches only the code needed to run a basic ice shelf test case.

This is a work-in-progress.

xylar committed:
Add Wright (1997) equation of state

A version of this equation is used in MOM6.  It has the
advantage that can be integrated analytically with pressure
(or depth with the Boussinesq approximation) to obtain
machine-precision pressure balance in the HGPF.

MiCurry committed:
Remove previous interpolation of greenfrac

MiCurry committed:
Remove previous interpolation of albedo12m

MiCurry committed:
Initialize and finalize the geotile manager for greenfrac

This commit initializes and finalizes the geotile manager for the climatology
monthly vegetation fraction (greenfrac). It currently does not do any
interpolation.

MiCurry committed:
Initialize and finalize geotile manager for albedo12m

This commit initializes and finalizes the geotile manager for the
climatological monthly albedo12m field. Currently interpolation is not done.

MiCurry committed:
Parallel interpolation of climatological monthly vegetation fraction

This commit adds logic to successfully interpolate climatological monthly
vegetation fraction (greenfrac) on a non-cvt partition decomposition. Because
the interpolation for the climatological monthly albedo field has not been
updated, the multiprocess check in init_atm_cases is still in place.

MiCurry committed:
Handle tile_z_start and tile_z_end dimensions in geographic index files

Some geographic datasets describe their z coordinate using tile_z_start and
tile_z_end. Before this commit, the geotile manager incorrectly assumed that
the z coordinate of geographic datasets was always specified as tile_z;
however, this is not the case as the climatological monthly albedo (albedo12m)
field lists its z coordinate as tile_z_start and tile_z_end in its index file.

Both tile_z and tile_z_start, tile_z_end methods are valid in the Description
of index Options section of Chapter 3 of the WRF User's Guide.

This commit updates the geotile manager to additionally handle tile_z_start,
tile_z_end.

MiCurry committed:
Parallel interpolation of climatological monthly albedo (albedo12m)

This commit adds logic to interpolate the climatological monthly albedo
(albedo12m) safely in parallel on a non-cvt partition. Although all fields now
have logic to safely interpolate static fields in parallel on a non-cvt
decomposition, the check in mpas_init_cases is still present so processing in
parallel is currently not possible.

MiCurry committed:
Remove multiprocess check for static interpolation

Now that all of the interpolations for all static fields have been updated to
safely be parallelized on non-cvz partition, the multiprocessor check for
static interpolation can be removed.

caozd999 committed:
Add prognostic-sediment-transport.rst

caozd999 committed:
Add prognostic sediment transport design doc, fix webpage bugs for rst.

caozd999 committed:
Fix format issues in the rst file

caozd999 committed:
Indent the unordered list two spaces inside one unordered list

caozd999 committed:
Text issues (typos, unit format, etc) cleanup

mgduda committed:
Remove unneeded uses of specific ESMF Fortran modules in MPAS infrastructure

The mpas_derived_types and mpas_timekeeping modules contained not only a 'use'
of the main ESMF module, but also uses of specific ESMF timekeeping modules
(ESMF_BaseMod, ESMF_Stubs, ESMF_CalendarMod, ESMF_ClockMod, ESMF_TimeMod, and
ESMF_TimeIntervalMod). Since all of the specific ESMF modules are ultimately
included in the main ESMF module, all that is needed in the mpas_derived_types
and mpas_timekeeping modules is a 'use ESMF' statement.

It's not clear why specific ESMF modules were being used originally, or why
commit dab101b5 didn't remove the use of specific ESMF modules when the use
of the main ESMF module was added.

caozd999 committed:
Remove sediment class component.

xylar committed:
Add namelist option for reference pressure for Wright

vanroekel committed:
Fixes shortwave absorption for shallow cells

Currently for very shallow cells some of the incident shortwave
radiation is not applied to the column.  This adds that missing heating
back to the bottom cell.

philipwjones committed:
fix for a PGI array syntax problem

This replaces array syntax in one loop with explicit loops to
work around a specific compiler issue with PGI.

philipwjones committed:
Cleanup of ocean split-explicit driver

This is a significant cleanup of the ocean split-explicit time
integration driver to prepare for future GPU and other performance
changes. Modifications include:

    - removal of block loops
    - removal of pointer retrievals for variables accessed
      through modules (config, mesh, diagnostics)
    - consolidation of remaining pointer retrievals
    - fixed misaligned indents and consistent indentation throughout
    - moved some static flags/masks to initialization and simplified
      some logic
    - eliminated a scratch pool variable and replaced with local allocate
    - reduced line lengths for better readability
    - changed some masks to reals to avoid unnecessary type conversion

A change was also made to the mode_forward driver to manage
different time integration init calls.

mgduda committed:
Merge branch 'framework/unneeded_esmf_mods' into develop (PR #797)

This merge removes unneeded uses of specific ESMF Fortran modules in MPAS
infrastructure.

The mpas_derived_types and mpas_timekeeping modules contained not only a 'use'
of the main ESMF module, but also uses of specific ESMF timekeeping modules
(ESMF_BaseMod, ESMF_Stubs, ESMF_CalendarMod, ESMF_ClockMod, ESMF_TimeMod, and
ESMF_TimeIntervalMod). Since all of the specific ESMF modules are ultimately
included in the main ESMF module, all that is needed in the mpas_derived_types
and mpas_timekeeping modules is a 'use ESMF' statement.

It's not clear why specific ESMF modules were being used originally, or why
commit dab101b didn't remove the use of specific ESMF modules when the use
of the main ESMF module was added.

* framework/unneeded_esmf_mods:
  Remove unneeded uses of specific ESMF Fortran modules in MPAS infrastructure

mgduda committed:
Merge branch 'init_atmosphere/par_greenfrac_12malb' into develop (PR #764)

This merge allows the init_atmosphere model to safely interpolate monthly
vegetation fraction (greenfrac) and monthly climatological albedo (albedo12m)
in parallel on non-cvt partitions.

Also included in this merge is the removal of the check that the number of
tasks was 1 when processing static fields. Now, all static fields can be
interpolated in parallel with an arbitrary mesh partition file.

* init_atmosphere/par_greenfrac_12malb:
  Remove multiprocess check for static interpolation
  Parallel interpolation of climatological monthly albedo (albedo12m)
  Handle tile_z_start and tile_z_end dimensions in geographic index files
  Parallel interpolation of climatological monthly vegetation fraction
  Initialize and finalize geotile manager for albedo12m
  Initialize and finalize the geotile manager for greenfrac
  Remove previous interpolation of albedo12m
  Remove previous interpolation of greenfrac

xylar committed:
Simplify computing landIcePressure and vert grid

While we supported 3 options for initializing the ice draft (SSH)
and landIcePressure, we now only use the simplest one, which uses
the denisty in the top layer as a guess for the effective density
of the ocean.

This merge significantly simplifies and clarifies the process for
creating the vertical coordinate and later initializing
landIcePressure in several test cases.  The subroutine for creating
the vertical coordinate ocn_init_vertical_grid() has been moved to
ocn_init_vertical_grids, where it clearly belongs.

xylar committed:
Remove config_iterative_init_variable

This config option is no longer needed because we always initialize
landIcePressure (not SSH) and we always use the density in the top
layer.

xylar committed:
Rename modifySSHMask --> modifyLandIcePressureMask

The new name is an accurate description of what it is used for,
now that we only allow modificiation of landIcePressure based on
SSH, and not the other way around.

xylar committed:
Update the 4 test cases with ice-shelf cavities

Each has been modified go use the new modifyLandIcePressureMask
variable and to call the more simplified ocn_init_vertical_grid()
function.

xylar committed:
Remove unused scratch variables

xylar committed:
Add minLevelCell variable

For now, only for use in init mode

xylar committed:
Add minLevelCell and z-level coord to init mode

cbegeman committed:
Add minLevelCell,Edge,Vertex variables to ocn_init_routines

cbegeman committed:
Add minLevelEdge,minLevelVertex variables to the Registry

xylar committed:
Compute minLevelCell and maxLevelCell for z-level

There's no need for the test case to figure these out.

vanroekel committed:
Fixes missing thread private

vanroekel committed:
Adds maxLevelCell check on fluxRemaining

cbegeman committed:
Add minLevel* variables to ocn_mesh

cbegeman committed:
Test k-loop range minLevelCell,maxLevelCell

cbegeman committed:
MinLevel* changes to ocn_validate_state:

First test of k-loop range minLevelCell,maxLevelCell

xylar committed:
Fix typo

Co-authored-by: Carolyn Begeman <cbegeman@lanl.gov>

xylar committed:
Add fill values above ice shelf in z-level test

cbegeman commented:
This is a first stab at dividing up the work. Feel free to edit or suggest changes if what I've suggested doesn't suit. I roughly ordered the list by priority, where we make the changes to the parts of the code that cannot be disabled first. This basically means that the tendency terms are updated later. For the most part, there's only one person touching each module and one person for each of the tendency groups (velocity, tracers, thickness). 

- "[X] E3SM" indicates that the changes in the module/subroutine passed E3SM tests (details TBD)
- "[X] Standalone" indicates that the changes in the module/subroutine passed the baseline regression suite for MPAS-Ocean [here](https://github.com/xylar/compass/tree/remove_config_iterative_init_variable)
- The checkbox at the start of the line indicates that the changes have been incorporated into this PR. Ideally this wouldn't happen until after it passes the corresponding standalone and E3SM tests. Right now there are a few exceptions to this rule to get us started.

Note: `minLevel*` variables are available through `use ocn_mesh`.
Note: use this loop syntax for now (Option 1 in [this discussion](https://github.com/MPAS-Dev/MPAS-Model/pull/811#issuecomment-777691337)):
```
do k = minLevelCell(iEdge), maxLevelCell(iEdge) !mlc-do-k
end do !mlc-end-do
```
Note: This may still not be a complete list. 

Init:
- [X] ocn_init_sub_ice_shelf_2D: @xylar [X] E3SM
   - [X] ocn_init_setup_sub_ice_shelf_2D() [X] Standalone
- [X] ocn_init_ssh_and_landIcePressure: @xylar
   - [X] ocn_init_ssh_and_landIcePressure_balance() [X] Standalone
- [x] ocn_init_vertical_grids: @xylar [x] Standalone [X] E3SM
   - [X] ocn_init_vertical_grid() [X] Standalone
   - [x] ocn_compute_z_level_layerThickness()  [X] Standalone
   - [x] ocn_compute_zMid_from_layerThickness()  [X] Standalone
   - [x] ocn_compute_z_star_layerThickness()  [X] Standalone
   - [x] ocn_alter_bottomDepth_for_pbcs()  [X] Standalone
   - [x] ocn_alter_ssh_for_ptcs()  [X] Standalone
   - [x] ocn_compute_Haney_number() [X] Standalone
   - [ ] ~~ocn_init_vertical_grid_with_max_rx1() [ ] Standalone~~  Probably doesn't make sense with `minLevelCell > 1`
   - [ ] ~~constrain_rx1_layer() [ ] Standalone~~ Same as above

Forward:
- [X] ocn_init_routines: @cbegeman [X] E3SM
   - [X] ocn_init_routines_compute_min_max_level() [X] Standalone
   - [X] ocn_init_routines_vert_coord() [X] Standalone
   - [X] ocn_init_routines_block() [X] Standalone
- [X] ocn_forcing: @xylar [X] E3SM
   - [X] ocn_forcing_build_fraction_absorbed_array [X] Standalone
- [X] ocn_mesh: @cbegeman [ ] Standalone [ ] E3SM
   - [X] minLevelCell
   - [X] minLevelEdgeTop
   - [X] minLevelEdgeBot
   - [X] minLevelVertexTop
   - [X] minLevelVertexBot
   - [X] ocn_meshCreate()
   - [X] ocn_meshUpdateFields()
   - [X] ocn_meshDestroy()
- [X] ocn_time_integration_split: @cbegeman [X] E3SM
   - [X] ocn_time_integration_split_init()  [X] Standalone
   - [X] ocn_time_integrator_split()  [X] Standalone
      - [X] Outside large iteration loop  [X] Standalone
      - [X] Inside large iteration loop  [X] Standalone
         Note: nothing is done here when config_disable_thick_all_tend .and. config_disable_vel_all_tend .and. config_disable_tr_all_tend
- [X] ocn_diagnostics: @cbegeman [X] Standalone [X] E3SM
   - [X] ocn_relativeVorticity_circulation() [X] Standalone
   - [X] ocn_diagnostic_solve_layerThicknessEdge() [X] Standalone
   - [X] ocn_diagnostic_solve_vorticity()
   - [X] ocn_diagnostic_solve_richardson() [X] Standalone
   - [X] ocn_diagnostic_solve_surfaceLayer() [X] Standalone
   - [X] ocn_diagnostic_solve_vortVel() [X] Standalone
   - [X] ocn_diagnostic_solve_z_coordinates() [X] Standalone
   - [X] ocn_diagnostic_solve_pressure() [X] Standalone
   - [X] ocn_vert_transport_velocity_top()
   - [X] ocn_fuperp() [X] Standalone
   - [X] ocn_compute_land_ice_flux_input_fields()
   - [X] ocn_validate_state() [X] Standalone
- [X] ocn_thick_ale: @cbegeman [X] Standalone [X] E3SM
   - [X] ocn_ALE_thickness
- [X] ocn_vmix_cvmix: @cbegeman [X] Standalone [X] E3SM
   - [X] ocn_vmix_coefs_cvmix_build	 

- [x] ocn_thick_hadv: @xylar [X] Standalone [X] E3SM
   - [x] ocn_hadv_thick_tend (config_disable_thick_hadv)
- [x] ocn_thick_vadv: @xylar [X] Standalone [X] E3SM
   - [x] ocn_vadv_thick_tend (config_disable_thick_vadv)
- [x] ocn_thick_surface_flux: @xylar [X] Standalone [X] E3SM
   - [x] ocn_thick_surface_flux_tend (config_disable_thick_sflux)	 

- [ ] ~~ocn_tendency~~: @cbegeman
   - [ ] ~~ocn_tend_tracer~~ (config_disable_tr_all_tend) Belongs in Phase 2

- [x] ocn_tracer_advection: @xylar [X] Standalone [X] E3SM
   - [x] ocn_tracer_advection_tend (config_disable_tr_adv)
- [x] ocn_tracer_advection_mono: @xylar [X] Standalone  [X] E3SM
   - [x] ocn_tracer_advection_mono_tend (config_disable_tr_adv)

- [x] ocn_vel_forcing_surface_stress: @xylar  [X] Standalone [X] E3SM
   - [x] ocn_vel_forcing_surface_stress_tend (config_disable_vel_surface_stress)
- [x] ocn_vel_hadv_coriolis: @xylar  [X] Standalone [X] E3SM
   - [x] ocn_vel_hadv_coriolis_tend (config_disable_vel_coriolis)
- [x] ocn_vel_vadv: @xylar  [X] Standalone [X] E3SM
   - [x] ocn_vel_vadv_tend (config_disable_vel_vadv)
- [x] ocn_vel_pressure_grad: @xylar  [X] Standalone [X] E3SM
   - [x] ocn_vel_pressure_grad_tend (config_disable_vel_pgrad)

- [X] ocn_vmix: @cbegeman [X] Standalone [X] E3SM
   - [X] ocn_vmix_implicit
   - [X] ocn_vel_vmix_tend_implicit (config_disable_vel_vmix)
   - [X] ocn_tracer_vmix_tend_implicit (config_disable_tr_vmix)

- [X] ocn_vel_hmix_del2: @cbegeman [X] Standalone [X] E3SM
   - [X] ocn_vel_hmix_del2_tend (config_disable_vel_hmix)

xylar commented on src/core_ocean/Registry.xml:
```diff
@@ -1629,6 +1631,7 @@
 			<var name="layerThickness"/>
 			<var name="ssh"/>
 			<var name="maxLevelEdgeTop"/>
+			<var name="minLevelEdgeTop"/>
```
I think `minLevelEdgeBot` would be the corresponding variable to `maxLevelEdgeTop`.

xylar commented on src/core_ocean/shared/mpas_ocn_init_routines.F:
```diff
@@ -502,24 +553,27 @@ subroutine ocn_init_routines_vert_coord(domain)!{{{
-                    call mpas_log_write(' iCell: $i, maxLevelCell(iCell): $i, bottomDepth(iCell): $r, sum(h): $r', &
-                       intArgs=(/iCell, maxLevelCell(iCell) /), &
-                       realArgs=(/ bottomDepth(iCell),sum(layerThickness(1:maxLevelCell(iCell),iCell)) /) )
+                    call mpas_log_write(' iCell: $i, minLevelCell(iCell), maxLevelCell(iCell): $i, bottomDepth(iCell): $r, sum(h): $r', &
+                       intArgs=(/iCell, minLevelCell(iCell), maxLevelCell(iCell) /), &
+                       realArgs=(/ bottomDepth(iCell),sum(layerThickness(minLevelCell(iCell):maxLevelCell(iCell),iCell)) /) )
```
```suggestion
                    call mpas_log_write(' iCell: $i, minLevelCell(iCell), maxLevelCell(iCell): $i, $i, bottomDepth(iCell): $r, sum(h): $r', &
                       intArgs=(/iCell, minLevelCell(iCell), maxLevelCell(iCell) /), &
                       realArgs=(/ bottomDepth(iCell),sum(layerThickness(minLevelCell(iCell):maxLevelCell(iCell),iCell)) /) )
```

xylar commented on src/core_ocean/shared/mpas_ocn_init_routines.F:
```diff
@@ -502,24 +553,27 @@ subroutine ocn_init_routines_vert_coord(domain)!{{{
-                    call mpas_log_write(' iCell: $i, maxLevelCell(iCell): $i, bottomDepth(iCell): $r, sum(h): $r', &
-                       intArgs=(/iCell, maxLevelCell(iCell) /), &
-                       realArgs=(/ bottomDepth(iCell),sum(layerThickness(1:maxLevelCell(iCell),iCell)) /) )
+                    call mpas_log_write(' iCell: $i, minLevelCell(iCell), maxLevelCell(iCell): $i, bottomDepth(iCell): $r, sum(h): $r', &
+                       intArgs=(/iCell, minLevelCell(iCell), maxLevelCell(iCell) /), &
+                       realArgs=(/ bottomDepth(iCell),sum(layerThickness(minLevelCell(iCell):maxLevelCell(iCell),iCell)) /) )
```
```suggestion
                    call mpas_log_write(' iCell: $i, minLevelCell(iCell), maxLevelCell(iCell): $i, $i bottomDepth(iCell): $r, sum(h): $r', &
                       intArgs=(/iCell, minLevelCell(iCell), maxLevelCell(iCell) /), &
                       realArgs=(/ bottomDepth(iCell),sum(layerThickness(minLevelCell(iCell):maxLevelCell(iCell),iCell)) /) )
```

xylar commented on src/core_ocean/shared/mpas_ocn_mesh.F:
```diff
@@ -307,7 +312,17 @@ subroutine ocn_meshCreate(domain) !{{{
+                                  minLevelVertexTop)
+         call mpas_pool_get_array(meshPool, 'minLevelVertexBot', &
+                                  minLevelVertexBot)
+         call mpas_pool_get_array(meshPool, 'minLevelCell', &
```
```suggestion
         call mpas_pool_get_array(meshPool, 'maxLevelCell', &
```

xylar commented:
@cbegeman, I found a few small things to correct in your changes.  Nothing that should affect us proceeding but things that need to be fixed soonish.

Also, you will need to add @dengwirda and me as collaborators (with write or maybe triage permission) to your fork so we can update the check marks above.  Let me know if you want more instructions on how to do that.

xylar commented on src/core_ocean/shared/mpas_ocn_init_routines.F:
```diff
@@ -93,7 +93,7 @@ module ocn_init_routines
 !
 !-----------------------------------------------------------------------
-subroutine ocn_init_routines_compute_max_level(domain)!{{{
+subroutine ocn_init_routines_compute_min_max_level(domain)!{{{
```
The corresponding `end subroutine ocn_init_routines_compute_min_max_level` change is needed.

xylar commented:
The branch as it is doesn't compile for me because of the missing change above (`end subroutine ocn_init_routines_compute_min_max_level`) so that's a high priority to fix.

xylar commented:
There are 2 other places where the update `ocn_init_routines_compute_max_level` --> `ocn_init_routines_compute_min_max_level` is needed:
https://github.com/cbegeman/MPAS-Model/blob/ocean/develop/src/core_ocean/mode_analysis/mpas_ocn_analysis_mode.F#L115
and
https://github.com/cbegeman/MPAS-Model/blob/ocean/develop/src/core_ocean/shared/mpas_ocn_init_routines.F#L68

It should also be fixed in this comment:
https://github.com/cbegeman/MPAS-Model/blob/ocean/develop/src/core_ocean/shared/mpas_ocn_init_routines.F#L87

xylar commented:
The fix I list above in `ocn_meshCreate` is also needed in order for the COMPASS regression tests to pass.

xylar committed:
Add minLevel* to ocn_compute_Haney_number

cbegeman commented:
@xylar Thanks for catching those mistakes. I think I've made all the changes you've requested. I'll get in touch about testing it with the regression suite.

mark-petersen committed:
Merge PR #804 'vanroekel/ocean/fixShortwaveFlux' into ocean/develop

Fixes shortwave absorption for shallow cells #804

Currently for very shallow cells some of the incident shortwave
radiation is not applied to the column. This adds that missing heating
back to the bottom cell.

cbegeman committed:
Set inactive top cells to -1e34, ocn_init_routines_block

cbegeman committed:
MinLevel* changes to ocn_time_integration_split

cbegeman committed:
MinLevel* changes to ocn_diagnostic_solve (pick tracer,velocity surface values)

cbegeman committed:
MinLevel* changes to ocn_relativeVorticity_Circulation

cbegeman committed:
MinLevel* changes to ocn_diagnostic_solve_layerThicknessEdge

cbegeman committed:
MinLevel* changes to ocn_diagnostic_solve_vorticity:

Note that do-loop from 1:maxLevelVertexBot could not be
changed to minLevelVertexTop:maxLevelVertexBot without
introducing bfb differences in QU240 test cases and some
sub_ice_shelf_2d test cases.

cbegeman committed:
MinLevel* changes to ocn_diagnostic_solve_richardson

cbegeman committed:
MinLevel* changes to ocn_diagnostic_solve_surfaceLayer

cbegeman committed:
MinLevel* changes to ocn_diagnostic_solve_vortVel

cbegeman committed:
MinLevel* changes to ocn_diagnostic_solve_z_coordinates

cbegeman committed:
MinLevel* changes to ocn_diagnostic_solve_pressure

cbegeman committed:
MinLevel* changes to ocn_vert_transport_velocity_top

cbegeman committed:
MinLevel* changes to ocn_fuperp

xylar committed:
Fix full partial top cells

xylar commented:
I added test cases based on @mark-petersen's comment: https://github.com/MPAS-Dev/MPAS-Model/pull/811#issuecomment-780079421
If you check out (or fetch and hard reset) my compass branch https://github.com/xylar/compass/tree/add_z_level_sub_ice_shelf_2D
```
$ ./list_testcases.py | grep z_level
  190: -o ocean -c sub_ice_shelf_2D -r 5km -t z_level
  191: -o ocean -c sub_ice_shelf_2D -r 5km -t z_level_full_cells
  192: -o ocean -c sub_ice_shelf_2D -r 5km -t z_level_full_cells_const_S
```
There are 3 versions of the `z_level*` test case that are hopefully self-explanatory.  I have also disabled all tendencies by default for now in the `forward` run.

I also updated this branch because I had made a mistake in the definition of full top cells.  I'll add my commit for adding `minLevelCell` to `ocn_compute_Haney_number` in a moment, too.

One way to prioritize our work on the modules and subroutines would be to find out where the `z_level_full_cells_const_S/forward` step breaks.  I'll play with this now.

xylar commented:
Current error I'm seeing is:
```
Program received signal SIGFPE: Floating-point exception - erroneous arithmetic operation.

Backtrace for this error:
#0  0x7f3c526b020f in ???
#1  0x55f80b1e4404 in __ocn_time_integration_split_MOD_ocn_time_integration_split_init
	at /home/xylar/code/mpas-work/compass/add_z_level_sub_ice_shelf_2D/MPAS-Model/ocean/new-min-level-cell-feature/src/core_ocean/mode_forward/mpas_ocn_time_integration_split.F:2156
```
I saw that @cbegeman has `ocn_time_integration_split_init` checked off but I guess that commit hasn't been pushed to the branch yet?

xylar commented:
Regarding `ocn_init_vertical_grids`:

I won't be putting `minLevelCell` in 2 functions related to the vertical coordinate with Haney-number constraint.  That would be a huge can of worms and would take quite a bit of thinking.  For now, the only vertical coordinate that supports `minLevelCell > 1` is `z-level`, not the `haney-number` coordinate created by this code.  So I think it's fine.

## Standalone testing

I tested on my Ubuntu laptop with Gnu and on Anvil with Intel, and the `z_level_dev` regression suite passed against the baseline.  For reference, the baseline is with compass branch:
https://github.com/xylar/compass/tree/remove_config_iterative_init_variable
and MPAS-Model branch:
https://github.com/xylar/MPAS-Model/tree/ocean/simplify_ssh_land_ice_pressure_init

## E3SM testing

I ran:
```
PEM_Ln9.ne30_oECv3_ICG.A_WCYCL1850S.anvil_intel
PEM_Ln9.T62_oQU240.GMPAS-IAF.anvil_gnu
PET_Ln3.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.anvil_intel
PET_Ln9.ne30_oECv3_ICG.A_WCYCL1850S.anvil_gnu
PET_Ln9_P1024.ne30_oECv3_ICG.A_WCYCL1850S.anvil_gnu
PET_Ln9.T62_oQU240.GMPAS-IAF.anvil_gnu
PET_Ln9.T62_oQU240.GMPAS-IAF.anvil_intel
SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.anvil_gnu
SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.anvil_intel
```
The last 2 were compared against a baseline of today's `E3SM/master`.

I will not be recording my testing at this level of detail in the future, just differences from this protocol.

cbegeman committed:
MinLevel* changes to ocn_compute_land_ice_flux_input_fields

mark-petersen committed:
Merge PR #812 'xylar/ocean/simplify_ssh_land_ice_pressure_init' into ocean/develop

Simplify computing landIcePressure and vertical grid in ocean init mode #812

This merge affects only init mode. It modifies the main registry and
diagnostics variables to rename an init-mode variable. All other changes
are in the mode_init directory.

While we have supported 3 options for initializing the ice draft (SSH)
and landIcePressure, we now only use the simplest one, which uses the
density in the top layer as a guess for the effective density of the
ocean.

This merge significantly simplifies and clarifies the process for
creating the vertical coordinate and later initializing landIcePressure
in several test cases by getting rid of the unused options. The
subroutine for creating the vertical coordinate ocn_init_vertical_grid()
has been moved to ocn_init_vertical_grids (where it clearly belongs).

The namelist option config_iterative_init_variable is no longer needed
and has been removed because we always initialize
landIcePressure (not SSH) and we always use the density in the top
layer.

The init mode diagnostic variable modifySSHMask has been renamed
modifyLandIcePressureMask to be consistent with these changes. We now
only modify landIcePressure, not SSH.

The 4 test cases with ice-shelf cavities:

global_ocean
isomip
isomip_plus
sub_ice_shelf_2D

have been updated with these changes.

xylar committed:
add minLevelCell loops to ocn_forcing_build_fraction_absorbed_array

cbegeman committed:
MinLevel* changes to ocn_ALE_thickness

xylar commented:
## testing of `ocn_forcing`

### COMPASS

The `z_level_dev` regression test passed with Gnu compilers on my Ubuntu laptop and Intel on Anvil, with the baseline listed in my previous post.

### E3SM

The following tests passed:
```
PEM_Ln9.ne30_oECv3_ICG.A_WCYCL1850S.anvil_intel
PEM_Ln9.T62_oQU240.GMPAS-IAF.anvil_gnu
PET_Ln9.ne30_oECv3_ICG.A_WCYCL1850S.anvil_gnu
PET_Ln9_P1024.ne30_oECv3_ICG.A_WCYCL1850S.anvil_gnu
PET_Ln9.T62_oQU240.GMPAS-IAF.anvil_gnu
SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.anvil_gnu.G  # baseline from today's E3SM/master
SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.anvil_gnu.C
SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.anvil_intel.G  # baseline from today's E3SM/master
SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.anvil_intel.C
```

The following test failed because of a missing data file unrelated to this work (see https://github.com/E3SM-Project/E3SM/pull/4092#issuecomment-782009399):
```
PET_Ln3.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.anvil_intel
```

xylar commented:
```
do k = minLevelCell(iEdge), maxLevelCell(iEdge) !mlc-do-k
end do !mlc-end-do
```
My understanding from [Phil's comment](https://github.com/MPAS-Dev/MPAS-Model/pull/811#issuecomment-781464654)  on the design document is that he's saying we don't need these comments.  @cbegeman, if that's your reading of that comment, too, I suggest we stop including these from now on and maybe remove the existing ones before making the "real" PR.

xylar committed:
Add intel-mpi to top Makefile

xylar committed:
Add minLevel* to ocn_thick_hadv_tend

xylar committed:
Add minLevel* to ocn_thick_vadv_tend

xylar committed:
Add minLevel* to ocn_thick_surface_flux_tend

xylar commented:
## testing of `ocn_thick_hadv`, `*_vadv` and `*_surface_flux`

### COMPASS

The `z_level_dev` regression test passed with Gnu compilers on my Ubuntu laptop and Intel on Anvil, with the baseline listed in my previous posts.

### E3SM

The following tests passed:
```
PET_Ln3.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.anvil_intel
PEM_Ln9.ne30_oECv3_ICG.A_WCYCL1850S.anvil_intel
PEM_Ln9.T62_oQU240.GMPAS-IAF.anvil_gnu
PET_Ln9.ne30_oECv3_ICG.A_WCYCL1850S.anvil_gnu
PET_Ln9_P1024.ne30_oECv3_ICG.A_WCYCL1850S.anvil_gnu
PET_Ln9.T62_oQU240.GMPAS-IAF.anvil_gnu
SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.anvil_gnu.C
SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.anvil_intel.C
```

@jonbob staged the missing file for the first test, so it is working again!

xylar commented:
@cbegeman, please assign me some more modules to work on unless there aren't any.

cbegeman commented:
### Testing of ocn_init_routines, *_validate_state, *_time_integration_split

**COMPASS**

The z_level_dev regression test passed with Intel compilers on grizzly against the baseline.

Baseline: 
MPAS-Dev/compass/legacy, https://github.com/MPAS-Dev/compass/commit/9f3378b278c3afd1c64c261c772271575ac426fe
with cherry-picked 
`commit 79a2aa46143f4cbecc735ffd143a1c0fd72f1567  REMOVE LATER!!! Add z-leve dev. test suite`
from https://github.com/xylar/compass/tree/add_z_level_sub_ice_shelf_2D


**E3SM**

The following tests passed:
```
SMS_D.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-knl_intel.C
SMS_D.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-knl_gnu.C
PET_Ln9.T62_oQU240.GMPAS-IAF.cori-knl_intel
PEM_Ln9.T62_oQU240.GMPAS-IAF.cori-haswell_gnu
PET_Ln3.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.cori-haswell_intel
PET_Ln9.ne30_oECv3_ICG.A_WCYCL1850S.cori-knl_gnu
PEM_Ln9.ne30_oECv3_ICG.A_WCYCL1850S.cori-knl_intel
```

cbegeman committed:
MinLevel* changes to ocn_tracer_vmix_tend_implicit:

MinLevel* changes to tridiagonal_solve_mult
must accompany these changes

cbegeman committed:
Remove comments flagging do minLevel* loops

cbegeman committed:
MinLevel* changes to ocn_vmix_implicit

xylar commented:
@cbegeman, I noticed you force-pushed to this branch.  If we're sharing the branch, that's not safe to do.  It could mean changes get lost.  If you want to edit or clean up the branch history, we should switch to having me (and @dengwirda) do PRs to this branch instead.

xylar commented:
I just put back my 3 commits that got force-removed.

cbegeman committed:
MinLevel* changes to ocn_vel_hmix_del2_tend

xylar committed:
Add minLevel* to ocn_vel_forcing_surface_stress_tend

xylar committed:
add minLevel* to ocn_vel_hadv_coriolis_tend

xylar committed:
Add maxLevel* to ocn_vel_vadv_tend

xylar committed:
add minLevel* to ocn_vel_pressure_grad_tend

xylar commented:
## testing of `ocn_vel_hadv_coriolis`, `*_vadv`, `*_forcing_surface_stress`, and `*_pressure_grad`

### COMPASS

The `z_level_dev` regression test passed with Gnu compilers on my Ubuntu laptop and Intel on Anvil, with the baseline listed in my previous posts.

### E3SM

The following tests passed:
```
SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.anvil_gnu.C
SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.anvil_intel.C
PEM_Ln9.T62_oQU240.GMPAS-IAF.anvil_gnu
PEM_Ln9_P1024.ne30_oECv3_ICG.A_WCYCL1850S.anvil_intel
PET_Ln9.T62_oQU240.GMPAS-IAF.anvil_gnu
```

But these failed:
```
PET_Ln9_P1024.ne30_oECv3_ICG.A_WCYCL1850S.anvil_gnu
PET_Ln3.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.anvil_intel
```
I'm working on rolling things back to figure out why.

**Update:**
```
PET_Ln3.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.anvil_intel
```
is failing on `E3SM/master`.  I did a fresh clone to make sure.  Results are here:
```
/lcrc/group/e3sm/ac.xylar/scratch/anvil/PET_Ln3.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.anvil_intel.20210220_061822_7tihl6
```
but
```
PET_Ln9_P1024.ne30_oECv3_ICG.A_WCYCL1850S.anvil_gnu
```
is succeeding on `E3SM/master`, and failing on this branch so that one needs investigation.
This test is succeeding on 13c14fb1 but failing on a8cbcf69.  This is very odd because I'm quite sure the latter was part of my testing yesterday, and it succeeded.
I tried again with a8cbcf69 and it passed this time.  I really don't get it!

**Second Update:**
I switched to testing on Cori for those two tests.  Both
```
PET_Ln3.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.cori-haswell_intel
```
and
```
PET_Ln9_P1024.ne30_oECv3_ICG.A_WCYCL1850S.cori-knl_gnu
```
pass.

cbegeman commented:
> I just put back my 3 commits that got force-removed.

Thanks. Sorry about that!

xylar commented:
No worries.

xylar committed:
Add maxLevel* to ocn_tracer_advection_mono_tend

The call to this funciton from ocn_tracer_advection_tend is also
updated to include minLevel*.

cbegeman committed:
Minlevel* changes to ocn_vel_vmix_tend_implicit

xylar commented:
## testing of `ocn_tracer_advection` and `*_mono`

### COMPASS

The `z_level_dev` regression test passed with Gnu compilers on my Ubuntu laptop and Intel on Anvil, with the baseline listed in my previous posts.

### E3SM

The following tests passed:
```
SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.anvil_gnu.C
SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.anvil_intel.C
PET_Ln3.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.anvil_intel
PEM_Ln9.T62_oQU240.GMPAS-IAF.anvil_gnu
PET_Ln9.T62_oQU240.GMPAS-IAF.anvil_gnu
```

But these failed:
```
PEM_Ln9_P1024.ne30_oECv3_ICG.A_WCYCL1850S.anvil_intel
PET_Ln9_P1024.ne30_oECv3_ICG.A_WCYCL1850S.anvil_gnu
```

**Update:**
I rebased onto @cbegeman's latest changes and re-tested.  Both tests passed.  Again, I think the failures are something finicky with Anvil like node problems.

cbegeman commented:
### Testing of ocn_diagnostics routines

**COMPASS**

The z_level_dev regression test passed with Intel compilers on grizzly against the baseline.

Baseline: 
MPAS-Dev/compass/legacy, https://github.com/MPAS-Dev/compass/commit/9f3378b278c3afd1c64c261c772271575ac426fe
with cherry-picked 
`commit 79a2aa46143f4cbecc735ffd143a1c0fd72f1567  REMOVE LATER!!! Add z-leve dev. test suite`
from https://github.com/xylar/compass/tree/add_z_level_sub_ice_shelf_2D


**E3SM**

The following tests passed:
```
SMS_D.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-knl_intel.C
SMS_D.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-knl_gnu.C
PET_Ln9.T62_oQU240.GMPAS-IAF.cori-knl_intel
PEM_Ln9.T62_oQU240.GMPAS-IAF.cori-haswell_gnu
PET_Ln3.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.cori-haswell_intel
PET_Ln9.ne30_oECv3_ICG.A_WCYCL1850S.cori-knl_gnu
PEM_Ln9.ne30_oECv3_ICG.A_WCYCL1850S.cori-knl_intel
```

Note: These loop limits could not be changed without introducing bfb changes in the z_level_dev regression suite.
https://github.com/cbegeman/MPAS-Model/blob/44c35746b6cd8e9b6595fd672c5583e5e64636ad/src/core_ocean/shared/mpas_ocn_diagnostics.F#L679-L689

cbegeman committed:
Minlevel* changes to ocn_vmix_cvmix

cbegeman commented:
### Testing of ocn_time_integration_split, *_vmix, *_vmix_cvmix

**COMPASS**

The z_level_dev regression test passed with Intel compilers on grizzly against the baseline.

Baseline: 
MPAS-Dev/compass/legacy, https://github.com/MPAS-Dev/compass/commit/9f3378b278c3afd1c64c261c772271575ac426fe
with cherry-picked 
`commit 79a2aa46143f4cbecc735ffd143a1c0fd72f1567  REMOVE LATER!!! Add z-leve dev. test suite`
from https://github.com/xylar/compass/tree/add_z_level_sub_ice_shelf_2D


**E3SM**
https://github.com/E3SM-Project/E3SM/commit/b3a8a9eea1598e57226018fb38956e5e91f46a12.

The following tests passed:
```
SMS_D.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-knl_intel.C
SMS_D.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-knl_gnu.C
PET_Ln9.T62_oQU240.GMPAS-IAF.cori-knl_intel
PEM_Ln9.T62_oQU240.GMPAS-IAF.cori-haswell_gnu
PET_Ln3.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.cori-haswell_intel
PET_Ln9.ne30_oECv3_ICG.A_WCYCL1850S.cori-knl_gnu
PEM_Ln9.ne30_oECv3_ICG.A_WCYCL1850S.cori-knl_intel
```

Note: some of the k-limits in cvmix could not be changed to `minLevel*:maxLevel*` without introducing non-bit-for-bit changes in `z_level_dev` suite. This may require further debugging for the ice-shelf test cases with `minLevelCell` not equal to 1.

mark-petersen committed:
Correct OpenMP typo

mark-petersen committed:
Merge PR #313 'xylar/ocean/add_wright_1997_eos' into ocean/develop

Add Wright (1997) equation of state #313

This merge adds a new equation of state (EOS) from [Wright (1997)](https://doi.org/10.1175/1520-0426(1997)014%3C0735:AEOSFU%3E2.0.CO;2):
      * This formulation of the new EOS is linear with depth (using the Boussinesq approximation of pressure as `p = p_s - rho0*g*(z - ssh)`), which will allow computation of the full pressure in the horizontal pressure gradient force (HPGF) to be computed analytically.  This is in contrast to the more complex formulation of the reference pressure used in the current Jacket and McDougall (JM) EOS.
      * The EOS accounts for surface pressure (for in situ density, but not potential density) from atmosphere, sea ice, frazil and ice shelves, unlike the JM EOS.
      * Potential density is referenced to `p = config_eos_wright_ref_pressure` (which is zero by default), not in the middle of the top reference grid cell as in JM.

COMPASS changes that were previously in this PR will instead be added to the `compass` package when the time comes.

xylar commented:
> Note: some of the k-limits in cvmix could not be changed to minLevel*:maxLevel* without introducing non-bit-for-bit changes in z_level_dev suite. This may require further debugging for the ice-shelf test cases with minLevelCell not equal to 1.

Could you make a list of code lines where this was the case?  I'd like to investigate.

cbegeman commented:
> Note: some of the k-limits in cvmix could not be changed to minLevel*:maxLevel* without introducing non-bit-for-bit changes in z_level_dev suite. This may require further debugging for the ice-shelf test cases with minLevelCell not equal to 1.

> Could you make a list of code lines where this was the case? I'd like to investigate.

It's this section:
https://github.com/cbegeman/MPAS-Model/blob/7ed4c0bd6082e744139f4e662f71fa13927fca46/src/core_ocean/shared/mpas_ocn_vmix_cvmix.F#L367-L430

xylar commented:
Thanks, that does look complicated.  I'll play around with it tomorrow and see if I can make any headway.

cbegeman commented:
A little thought is needed to resolve a discrepancy between `minLevelCell`,`maxLevelCell` assignments in the sub-ice-shelf 2d setup and computing `minLevelEdge*`,`minLevelVertex*` in `ocn_init_routines_compute_min_max_level`:

The problem is that we have at least one cell where `maxLevelCell` = 0 (I presume indicating an empty water column), and `minLevelCell` = 1. I imagine, though haven't verified, that `minLevelCell` is set to 1 here and not subsequently overwritten.
https://github.com/cbegeman/MPAS-Model/blob/7ed4c0bd6082e744139f4e662f71fa13927fca46/src/core_ocean/mode_init/mpas_ocn_init_sub_ice_shelf_2D.F#L248

Then `minLevelVertexTop` is also set to 1 and `layerThicknessVertex` = 0 (due to using the wrong k-level in neighboring cells), causing `ocn_diagnostic_solve_vorticity` to fail. Oh, and I should note that this is after modifying the lower loop limit to `minLevelVertexTop` on the following line:
https://github.com/cbegeman/MPAS-Model/blob/7ed4c0bd6082e744139f4e662f71fa13927fca46/src/core_ocean/shared/mpas_ocn_diagnostics.F#L679
When I set the loop limits as such for the z_level_dev test suite it failed QU240 and sub-ice-shelf-2d tests; now I think the failure could be due to a similar issue.

My inclination is to set `minLevelCell` to `nVertLevels` when the water column is empty so it's kind of analogous to `maxLevelCell` being set to 0 when the water column is empty, but I haven't thoroughly vetted this. In looking into this, I also noticed these lines, which could be problematic:
https://github.com/cbegeman/MPAS-Model/blob/7ed4c0bd6082e744139f4e662f71fa13927fca46/src/core_ocean/mode_init/mpas_ocn_init_vertical_grids.F#L518-L519

Conceptually, I'd think that when `minLevelCell`>`maxLevelCell` the water column is empty and that when `minLevelCell`=`maxLevelCell`, the water column is one cell tall.

@xylar Any thoughts since you're more familiar with these test cases and the mode_init routines?

xylar commented:
> I imagine, though haven't verified, that minLevelCell is set to 1 here and not subsequently overwritten.

Yes, that's correct.

> My inclination is to set minLevelCell to nVertLevels when the water column is empty so it's kind of analogous to maxLevelCell being set to 0 when the water column is empty, but I haven't thoroughly vetted this.

I think that's nearly correct.  I think you probably want `minLevelCell = nVertLevels+1` (one-based indexing in Fortran) so `minLevelCell > maxLevelCell` even when `maxLevelCell = nVertLevels`.  I had this in the back of my mind, since this is exactly what I did in POP2x, the version of POP that had ice-shelf cavities.  But I hadn't noticed where this would be important.

I also don't think we need to set this in init mode.  I think we can modify `minLevelCell` to be `nVertLevels+1` in `ocn_init_routines_compute_min_max_level` whenever `maxLevelCell < minLevelCell`, which presumably will only happen when `maxLevelCell = 0`.

> In looking into this, I also noticed these lines, which could be problematic:

I don't see the problem with that line, other than the fact that `bottomDepth` is positive down and `ssh` is positive up, which is confusing.  That line says that the layer thickness is the water-column thickness.  In practice, I think the numerics of MPAS-Ocean don't handle situations where the water column is a single cell thick, but this is not the routine where that should be inforced so I think it remains correct.

Let's chat!

mattdturner committed:
Merge PR #806 'pwjones/ocean/pgiSyntaxFix' into ocean/develop

This replaces array syntax in one loop in the ocean tend_tracer routine
with explicit loops to
work around a compiler issue with PGI.

Fixes #805

vanroekel committed:
Fixes missing edgeSignOnCell in redi_term2

vanroekel committed:
Removes shortwave energy conservation fix by default

Coupled testing showed unexpected extremely strong sensitivity to the
previous shortwave radiation energy fixer.  In that fix the excess
energy was used to heat the bottom layer.  This yielded a strong
response in tropical precipitation, dramatically increasing the double
ITCZ bias.  This removes that energy fixer and only tracks the solar
radiation that is discarded for extremely shallow cells

vanroekel committed:
Adds option to enable shortwave energy fixer

off by default

cbegeman committed:
Do not use dry cells to determine minLevelVertex*

cbegeman committed:
Do not use dry cells to determine minLevelEdge*

mark-petersen committed:
Merge PR #820 'vanroekel/ocean/removeShortwaveEnergyConservationFixer' into ocean/develop

Removes shortwave energy conservation fix #820

Coupled testing showed unexpected extremely strong sensitivity to the
previous shortwave radiation energy fixer. In that fix the excess
energy was used to heat the bottom layer. This yielded a strong
response in tropical precipitation, dramatically increasing the double
ITCZ bias. This removes that energy fixer and only tracks the solar
radiation that is discarded for extremely shallow cells

mark-petersen committed:
Merge PR #819 'vanroekel/ocean/fixRediTerm2bug' into ocean/develop

Fixes missing edgeSignOnCell in redi_term2 #819

Currently redi_term2 is missing a factor of edgeSignOnCell in its
calculation. It is being correctly applied in the limiter, but not the
original calculation.

xylar committed:
Merge branch 'xylar/framework/add_intel_mpi' into develop

Builds with Intel MPI compilers require different names for FC_PARALLEL,
etc. This merge adds intel-mpi as one of the make targets in the
top-level Makefile.

* xylar/MPAS-Model/framework/add_intel_mpi:
  Add intel-mpi to top Makefile

cbegeman committed:
Check for layerThicknessVertex=0 in vorticity solve

mark-petersen committed:
Merge 'develop' into ocean/develop

* origin/develop:
  Add intel-mpi to top Makefile
  Remove multiprocess check for static interpolation
  Parallel interpolation of climatological monthly albedo (albedo12m)
  Handle tile_z_start and tile_z_end dimensions in geographic index files
  Remove unneeded uses of specific ESMF Fortran modules in MPAS infrastructure
  Parallel interpolation of climatological monthly vegetation fraction
  Initialize and finalize geotile manager for albedo12m
  Initialize and finalize the geotile manager for greenfrac
  Remove previous interpolation of albedo12m
  Remove previous interpolation of greenfrac

cbegeman commented:
### Testing `z_level_full_cells_const_S_linear_eos`
compass: https://github.com/xylar/compass/tree/add_z_level_sub_ice_shelf_2D at commit 194ee2c 

MPAS-Model: https://github.com/cbegeman/min-level-cell-test at commit 5a65880d
- Rebased onto 3744f94c `(xylar/MPAS-Model/ocean/add_z_level_sub_ice_shelf_2D) Fix top density in landIcePressure computation`

Unless otherwise noted, `config_disable_* = .true.`
Unless otherwise noted, `config_eos_type = 'linear'`. JM EOS is the standard but the resulting density field is non-uniform and thus less convenient for testing tracer tendencies.

- [X] Thickness tendency terms: `config_disable_thick_* = .false.` and config_eos_type = 'jm'`
   - [X] Passes `check_z_level_bit_for_bit.py`
- [ ] Velocity tendency terms: `config_disable_vel_* = .false.`
   - [X] `check_z_level_bit_for_bit.py`: same level of velocity errors as when only `pgrad` is enabled
   - [ ] compare directly with case when only `pgrad` is enabled
   - [X] `config_disable_vel_all_tend = .false.`,
   `config_disable_vel_pgrad = .false.`
      - [X] `check_z_level_bit_for_bit.py`
         - `normalVelocity` not bfb due to horizontal pressure gradient force not perfectly balancing coriolis force in regions with high gradients in the depth of `minLevelCell` (threshold appears to be dz/dx ~= 60m/5km for linear EOS.
         - JM EOS also tested and yields slightly smaller `normalVelocity` fluctuations. 
- [ ] Tracer tendency terms: `config_disable_tr_* = .false.`
   - Currently segfault-ing
   - [ ] Passes `check_z_level_bit_for_bit.py`
   - [ ] `config_disable_tr_all_tend = .false.`
      - [ ] Passes `check_z_level_bit_for_bit.py`: normalVelocity not bfb

xylar commented:
> Passes check_z_level_bit_for_bit.py: Currently segfault-ing

I assume the segfault is in the code, not in the python code for checking (which shouldn't even be able to segfault)?

cbegeman commented:
> I assume the segfault is in the code, not in the python code for checking (which shouldn't even be able to segfault)?

Right. The way I had written it was confusing. Now edited above.

mark-petersen committed:
Change default output to single precision

xylar committed:
Fix top density in landIcePressure computation

It is now computed with minLevelCell, not 1

mark-petersen committed:
Merge PR #414 'ocean/single_precision_output' into ocean/develop

Change output to single precision in stand-alone default streams files #414

This adds the flag

precision="single"

to the default streams file for output files. It does not alter restart
streams. Bit-for-bit testing in the nightly regression remains 8-byte
reals because that is specified in compass.

The purpose of adding single precision to the default streams is to
remind users that single precision is sufficient and would typically be
used for output. This does not affect E3SM, as the streams files are
created from a script there. We are changing to single precision
analysis output there as well, see E3SM-Project/E3SM#3360

vanroekel committed:
Simplifies the shortwave fixer formulation

makes the shortwave fixer add energy at bottom of column instead of
spreading throughout.  This is more physically correct

mattdturner committed:
Initialize btrvel_temp to 0 after allocate

mattdturner committed:
Remove btrvel_temp from Registry

philipwjones committed:
Merge branch 'ocean/develop' into ocean/sedriver

mark-petersen committed:
Merge PR #822 'vanroekel/ocean/improveShortwaveEnergyConservationFixer' into ocean/develop

Simplifies the shortwave fixer formulation #822

makes the shortwave fixer add energy at bottom of column instead of
spreading throughout. This is more physically correct

mark-petersen committed:
Correct typos and enabled to enable in flag

mark-petersen committed:
Remove trailing white spaces

mark-petersen committed:
Merge PR #793 'caozd999/ocean/sediment_transport_design_doc' into ocean/develop

Ocean/sediment transport design doc #793

This PR adds the prognostic sediment transport design document. It would
also be used for discussions about where and how

to define the sediment transport related variables (have been/will
be listed in the design document)
to implement each development requirement

This design document will keep updated along with the code
development work.

mark-petersen committed:
Merge PR #781 'philipwjones/ocean/sedriver' into ocean/develop

cleanup of ocean split-explicit driver #781

This is a significant cleanup of the ocean split-explicit time integration driver to prepare for future GPU and other performance changes. Modifications include:

   - removal of block loops
   - removal of pointer retrievals for variables accessed
     through modules (config, mesh, diagnostics)
   - consolidation of remaining pointer retrievals
   - fixed misaligned indents and consistent indentation throughout
   - moved some static flags/masks to initialization and simplified
     some logic
   - eliminated a scratch pool variable and replaced with local allocate
   - reduced line lengths for better readability
   - changed some masks to reals to avoid unnecessary type conversion

Bit-for-bit on Summit with PGI and QU240. I expect it to be bit-for-bit on other machines/configurations too.

xylar committed:
Unstretch z-level layerThickness to get restingThickness

cbegeman committed:
Change k-limits on kineticEnergyCell computation

cbegeman committed:
Set vertical velocity top to 0 from k=1:minLevelCell-1
